server/etcdserver/api/v3discovery/discovery.go

```diff
@@ -0,0 +1,569 @@
+// Copyright 2022 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package v3discovery provides an implementation of the cluster discovery that
+// is used by etcd with v3 client.
+package v3discovery
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+
+	"math"
+	"net/url"
+	"path"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"go.etcd.io/etcd/client/pkg/v3/transport"
+	"go.etcd.io/etcd/client/pkg/v3/types"
+	"go.etcd.io/etcd/client/v3"
+
+	"github.com/jonboulle/clockwork"
+	"go.uber.org/zap"
+)
+
+const (
+	discoveryPrefix = "/_etcd/registry"
+)
+
+var (
+	ErrInvalidURL     = errors.New("discovery: invalid peer URL")
+	ErrBadSizeKey     = errors.New("discovery: size key is bad")
+	ErrSizeNotFound   = errors.New("discovery: size key not found")
+	ErrFullCluster    = errors.New("discovery: cluster is full")
+	ErrTooManyRetries = errors.New("discovery: too many retries")
+)
+
+var (
+	// Number of retries discovery will attempt before giving up and error out.
+	nRetries             = uint(math.MaxUint32)
+	maxExpoentialRetries = uint(8)
+)
+
+type DiscoveryConfig struct {
```

Comment:
This config is part of embed.Config, As so we should annotate its fields with `json`. Can you replace flags names with `json` annotations?

Reply:
Done & thanks.

Reply:
Independently how we structure the flags (the other comment), 
I think we should structure this configuration to be explicit what's just `client` config, 
and what's `discovery` specific (in this case - nothing). 

How about structuring this protos/json as follows (ie. grouping `client` stuff in the `client:` submessage. 

```
type ClientConfig struct {  // Ideally in `client/v3/...`
  Conn string `json:"conn"`
  DialTimeout time.Duration `json:"dial-timeout"`
  ... 
}

type DiscoveryConfig struct {
  Client client `json:"client"`
}
```

Reply:
Yes, it's a good point. Actually previously @serathius already raised a [similar comment](https://github.com/etcd-io/etcd/pull/13635#discussion_r792504627). 

Since it may have some impacts on the clientv3 and also the etcdctl tool, so I'd like to resolve this comment in a separate PR. 

There is already a [clientconfig](https://github.com/etcd-io/etcd/blob/main/etcdctl/ctlv3/command/global.go#L100), but it's in etcdctl instead of clientv3.  So we may need some minor refactor. So I prefer to doing it in a separate PR. 

What do you think?  @ptabor @serathius @spzala